### PR TITLE
JAVA-2682: Create a Quarkus Cassandra extension

### PR DESCRIPTION
**Pull request by tomekl007**
It does not have support for:
- native
- health checks
- metrics
Those will be implemented as separate tasks (separate PRs)

### CODE REVIEW

File: bom/runtime/pom.xml [hunk 1]
```diff
@@ -191,6 +192,7 @@
         <error-prone-annotations.version>2.2.0</error-prone-annotations.version>
         <jib-core.version>0.13.0</jib-core.version>
         <google-http-client.version>1.34.0</google-http-client.version>
+
```
Comment (adutra) [inline bom/runtime/pom.xml]: Be careful with unnecessary changes like this when contributing to other repos.

File: bom/runtime/pom.xml [hunk 2]
```diff
@@ -863,6 +865,39 @@
                 <version>${project.version}</version>
             </dependency>
 
+            <dependency>
+                <groupId>com.datastax.oss</groupId>
+                <artifactId>java-driver-core</artifactId>
+                <version>${datastax.java.driver.version}</version>
+                <exclusions>
```
Comment (adutra) [inline bom/runtime/pom.xml:873]: Why do we need to exclude Tinkerpop?

File: bom/runtime/pom.xml [hunk 3]
```diff
@@ -863,6 +865,39 @@
...
+                    </exclusion>
+                    <exclusion>
+                        <groupId>org.apache.tinkerpop</groupId>
+                        <artifactId>gremlin-driver</artifactId>
+                    </exclusion>
+                    <exclusion>
+                        <groupId>org.apache.tinkerpop</groupId>
+                        <artifactId>tinkergraph-gremlin</artifactId>
+                    </exclusion>
+                    <exclusion>
```
Comment (adutra) [inline bom/runtime/pom.xml:890]: Is this because ASM doesn't work with Graal?  I see that ASM is only used by `com.github.jnr:jnr-ffi:jar:2.1.10:compile` and according to JAVA-2616, jnr-ffi is not working anyways with Graal, so isn't it simpler to exclude `jnr-ffi` ?

File: bom/runtime/pom.xml [hunk 4]
```diff
@@ -863,6 +865,39 @@
...
+                        <artifactId>asm-util</artifactId>
+                    </exclusion>
+                </exclusions>
+            </dependency>
+            <dependency>
+                <groupId>com.datastax.oss</groupId>
+                <artifactId>java-driver-mapper-runtime</artifactId>
+                <version>${datastax.java.driver.version}</version>
+            </dependency>
+            <dependency>
```
Comment (adutra) [inline bom/runtime/pom.xml:901]: Do we really need this one at runtime? This is typically only needed at compile-time (to generate code).

File: bom/runtime/pom.xml [hunk 5]
```diff
@@ -163,6 +163,7 @@
         <proton-j.version>0.33.3</proton-j.version>
         <okhttp.version>3.14.6</okhttp.version>
         <sentry.version>1.7.30</sentry.version>
+        <datastax.java.driver.version>4.5.0</datastax.java.driver.version>
```
Comment (adutra) [inline bom/runtime/pom.xml]: Looking at other properties, I think we should rename this to `datastax-java-driver.version`.

**Review by adutra — COMMENTED** [review]
I just did a first stab at this PR, will continue tomorrow. Thanks!

File: bom/runtime/pom.xml (hunk 2 above)
Comment (absurdfarce) [reply bom/runtime/pom.xml:873]: Presumably the concern here was along the lines of what was discussed in KAF-189?  If that's the case I don't think we need to worry about it here.  The underlying issue in KAF-189 was a packaging issue with the sink uberjar; it shouldn't apply to the core JAR.  Even if were a problem with the core JAR that would seem to be an argument to fix the packaging issue rather than to exclude Tinkerpop all together.

I say that because we need Tinkerpop if we have any intent of supporting graph queries within this extension.  I guess I'm assuming that is a goal, but if that's correct it's hard to see how we get away from Tinkerpop here.

File: bom/runtime/pom.xml (hunk 3 above)
Comment (absurdfarce) [reply bom/runtime/pom.xml:890]: Nit: jnr-ffi can be made to work with _Graal_, but something else in _Quarkus_ is causing it not to work correctly.

Also note that if we exclude jnr-ffi we have no chance of com.datastax.oss.driver.internal.core.os.Native working correctly so we should explicitly force the Java clock in that case.

File: extensions/cassandra-client/runtime/src/main/java/io/quarkus/cassandra/runtime/AbstractCassandraClientProducer.java [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+                DefaultDriverConfigLoader.DEFAULT_ROOT_PATH) {
+            @NonNull
+            @Override
+            public DriverConfigLoader build() {
+                return new NonReloadableDriverConfigLoader(super.build());
+            }
+        };
+    }
+
+    CassandraClientConfig getCassandraClientConfig() {
```
Comment (olim7t) [inline extensions/cassandra-client/runtime/src/main/java/io/quarkus/cassandra/runtime/AbstractCassandraClientProducer.java:49]: Nit: the corresponding setter is named `setConfig`

File: extensions/cassandra-client/deployment/src/main/java/io/quarkus/cassandra/deployment/CassandraClientProcessor.java [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+    }
+
+    private String getCassandraClientProducerClassName() {
+        return AbstractCassandraClientProducer.class.getPackage().getName() + "."
+                + "CassandraClientProducer";
+    }
+
+    private void createCassandraClientProducerBean(
+            BuildProducer<GeneratedBeanBuildItem> generatedBean,
+            String cassandraClientProducerClassName) {
```
Comment (olim7t) [inline extensions/cassandra-client/deployment/src/main/java/io/quarkus/cassandra/deployment/CassandraClientProcessor.java:63]: This method generates the following class at runtime:
```java
@ApplicationScoped
public class CassandraClientProducer extends AbstractCassandraClientProducer {

  @ApplicationScoped
  @Produces
  @Default
  @Unremovable
  public CqlSession createDefaultCassandraClient() {
    return createCassandraClient(this.getCassandraClientConfig());
  }
}
```
Why? There's nothing dynamic in there, the class could be defined directly in our sources.

File: extensions/cassandra-client/runtime/src/main/java/io/quarkus/cassandra/runtime/AbstractCassandraClientProducer.java [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+public abstract class AbstractCassandraClientProducer {
+
+    private CassandraClientConfig config;
+
+    public void setConfig(CassandraClientConfig config) {
+        this.config = config;
+    }
+
+    private ProgrammaticDriverConfigLoaderBuilder createDriverConfigLoader() {
+        return new DefaultProgrammaticDriverConfigLoaderBuilder(
```
Comment (olim7t) [inline extensions/cassandra-client/runtime/src/main/java/io/quarkus/cassandra/runtime/AbstractCassandraClientProducer.java:30]: Maybe we should completely drop Typesafe config, it sounds a bit at odds with Quarkus' native configuration mechanisms.

That should be easier once JAVA-2600 is merged (which I plan to prioritize for 4.6.0). Then it would be nice to have some sort of mapping from Quarkus options to driver options (based on some common prefix), without listing them explicitly in config classes.

**Review by olim7t — COMMENTED** [review]
Preliminary remarks, I'm still wrapping my head around how this all works.

File: bom/runtime/pom.xml (hunk 2 above)
Comment (tomekl007) [reply bom/runtime/pom.xml:873]: Are exclusions that I did are necessary because of the `<bannedDependencies>` specified here:
https://github.com/riptano/quarkus/blob/master/build-parent/pom.xml#L218

File: bom/runtime/pom.xml (hunk 3 above)
Comment (tomekl007) [reply bom/runtime/pom.xml:890]: I will be investigating that more as part of https://datastax-oss.atlassian.net/browse/JAVA-2696. The note about `TIMESTAMP_GENERATOR_FORCE_JAVA_CLOCK` is in comments

File: bom/runtime/pom.xml (hunk 4 above)
Comment (tomekl007) [reply bom/runtime/pom.xml:901]: This is used only in `integration-tests` module and somehow integration tests module has access to dependencies that are in this `runtime/pom.xml`

File: extensions/cassandra-client/deployment/src/main/java/io/quarkus/cassandra/deployment/CassandraClientProcessor.java (hunk 7 above)
Comment (tomekl007) [reply extensions/cassandra-client/deployment/src/main/java/io/quarkus/cassandra/deployment/CassandraClientProcessor.java:63]: I was following the pattern used for `mongo-client` in `MongoClientProcessor` and they are also using auto-generation. Right now it is the simplest `buildItem` but probably it will evolve and we would need to do the generation of this code so I propose to keep it this way right now (at least until all needed tasks will be done) then we will see if this code can be simplified, ok?

File: extensions/cassandra-client/runtime/src/main/java/io/quarkus/cassandra/runtime/AbstractCassandraClientProducer.java (hunk 8 above)
Comment (tomekl007) [reply extensions/cassandra-client/runtime/src/main/java/io/quarkus/cassandra/runtime/AbstractCassandraClientProducer.java:30]: Which part of the type-safe config do you want to drop?
I think that `DefaultProgrammaticDriverConfigLoaderBuilder` and fallback strategies are nice because allow clients to pass `application.conf` or `application.json` with java driver underlying settings without porting them all to quarkus mechanism. On the other hand, we can chain it with `ProgrammaticDriverConfigLoaderBuilder` that will fetch the required settings from `CassandraClientConnectionConfig` One note: Quarkus is doing a compile-time check of settings from `application.properties` so the map-based config will not be a good fit here.

File: extensions/cassandra-client/deployment/src/main/java/io/quarkus/cassandra/deployment/CassandraClientProcessor.java [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+    }
+
+    @BuildStep
+    HealthBuildItem addHealthCheck(CassandraClientBuildTimeConfig buildTimeConfig) {
+        return new HealthBuildItem("io.quarkus.cassandra.runtime.health.CassandraHealthCheck",
+                buildTimeConfig.healthEnabled, "cassandra");
+    }
+
+    @BuildStep
+    RuntimeInitializedClassBuildItem runtimeMetadataManager() {
```
Comment (tomekl007) [inline extensions/cassandra-client/deployment/src/main/java/io/quarkus/cassandra/deployment/CassandraClientProcessor.java:120]: `MetadataManager` must be initialized at runtime because it uses `Inet4Socket` address that cannot initialized at deployment time because of:
```
No instances of java.net.Inet4Address are allowed in the image heap as this class should be initialized at image runtime.
````

File: extensions/cassandra-client/runtime/src/main/java/io/quarkus/cassandra/runtime/AbstractCassandraClientProducer.java (hunk 8 above)
Comment (olim7t) [reply extensions/cassandra-client/runtime/src/main/java/io/quarkus/cassandra/runtime/AbstractCassandraClientProducer.java:30]: I want to drop the dependency to `com.typesafe:config`, and then inject the Quarkus config into the map-based config loader introduced by JAVA-2600.

> fallback strategies are nice because allow clients to pass application.conf or application.json with java driver underlying settings without porting them all to quarkus mechanism

`OptionsMap` can also be made to initialize with the driver defaults, so you don't have to pass every option.

File: extensions/cassandra-client/deployment/src/main/java/io/quarkus/cassandra/deployment/CassandraClientProcessor.java (hunk 7 above)
Comment (olim7t) [reply extensions/cassandra-client/deployment/src/main/java/io/quarkus/cassandra/deployment/CassandraClientProcessor.java:63]: How about going the other way: start simple, and add complexity only when we need it.

I spent a lot of time figuring out this method, it's even harder to understand if it does something that's not strictly necessary.

Comment (tomekl007) [reply extensions/cassandra-client/deployment/src/main/java/io/quarkus/cassandra/deployment/CassandraClientProcessor.java:63]: I've tried to simplify this as you suggested.
Unfortunately, there are some problems with order of initialization, and this:
```
 public RuntimeValue<CqlSession> getClient() {
    return new RuntimeValue<>(Arc.container().instance(CqlSession.class, defaultName()).get());
  }
```
is failing with null pointer, because there is no `CqlSession` yet, I'll try to investigate more

**Comment by tomekl007** [comment]
I am closing this PR in favor of: 
https://github.com/datastax/cassandra-quarkus/pull/1
that is in our dedicated repo. 
@olim7t @adutra @absurdfarce please continue there.

**Comment by adutra** [comment]
The link above is not correct, the new PR is https://github.com/datastax/cassandra-quarkus/pull/1.
